File: app/src/main/java/com/example/paetz/yacguide/AscendActivity.java
package com.example.paetz.yacguide;

import android.app.Activity;
import android.app.DatePickerDialog;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.Spinner;

import com.example.paetz.yacguide.database.AppDatabase;
import com.example.paetz.yacguide.database.Ascend;
import com.example.paetz.yacguide.database.Partner;
import com.example.paetz.yacguide.database.Route;
import com.example.paetz.yacguide.utils.AscendStyle;
import com.example.paetz.yacguide.utils.IntentConstants;

import java.util.ArrayList;
import java.util.Calendar;

public class AscendActivity extends AppCompatActivity {

    private AppDatabase _db;
    private Ascend _ascend;
    private Route _route;
    private ArrayList<Integer> _partnerIds;
    private int _resultUpdated;
    private int _styleId;
    private int _year;
    private int _month;
    private int _day;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_ascend);

        _db = MainActivity.database;
        _resultUpdated = IntentConstants.RESULT_NO_UPDATE;
        _ascend = _db.ascendDao().getAscend(getIntent().getIntExtra(IntentConstants.ASCEND_KEY, AppDatabase.INVALID_ID));
        final int routeId = getIntent().getIntExtra(IntentConstants.ROUTE_KEY, AppDatabase.INVALID_ID);
        _route = _db.routeDao().getRoute(routeId == AppDatabase.INVALID_ID ? _ascend.getRouteId() : routeId);
        // Beware: _route may still be null (if the route of this ascend has been deleted meanwhile)
        _partnerIds = getIntent().getIntegerArrayListExtra(IntentConstants.ASCEND_PARTNER_IDS);
        if (_partnerIds == null) {
            _partnerIds = (_ascend == null) ? new ArrayList<Integer>() : _ascend.getPartnerIds();
        }
        _styleId = _year = _month = _day = 0;

        findViewById(R.id.notesEditText).setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                final InputMethodManager imm = (InputMethodManager) v.getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
                imm.hideSoftInputFromWindow(v.getWindowToken(), 0);
            }
        });

        _displayContent();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK) {
            _partnerIds = data.getIntegerArrayListExtra(IntentConstants.ASCEND_PARTNER_IDS);
            _displayContent();
        }
    }

    public void enter(View v) {
        Ascend ascend = new Ascend();
        if (_ascend != null) {
            ascend.setId(_ascend.getId());
        } else {
            _db.routeDao().updateAscendCount(_db.routeDao().getAscendCount(_route.getId()) + 1, _route.getId());
            _db.rockDao().updateAscended(true, _db.routeDao().getRoute(_route.getId()).getParentId());
            _resultUpdated = IntentConstants.RESULT_UPDATED;
        }
        ascend.setRouteId(_route == null ? _ascend.getRouteId() : _route.getId());
        ascend.setStyleId(_styleId);
        ascend.setYear(_year);
        ascend.setMonth(_month);
        ascend.setDay(_day);
        ascend.setPartnerIds(_partnerIds);
        ascend.setNotes(((EditText) findViewById(R.id.notesEditText)).getText().toString());
        _db.ascendDao().insert(ascend);
        Intent resultIntent = new Intent();
        setResult(_resultUpdated, resultIntent);
        finish();
    }

    public void cancel(View v) {
        Intent resultIntent = new Intent();
        setResult(IntentConstants.RESULT_NO_UPDATE, resultIntent);
        finish();
    }

    public void enterDate(View v) {
        final Calendar calendar = Calendar.getInstance();
        final int yy = calendar.get(Calendar.YEAR);
        final int mm = calendar.get(Calendar.MONTH);
        final int dd = calendar.get(Calendar.DAY_OF_MONTH);
        DatePickerDialog datePicker = new DatePickerDialog(AscendActivity.this, new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
                _year = year;
                _month = monthOfYear + 1;
                _day = dayOfMonth;
                ((EditText) findViewById(R.id.dateEditText)).setText(_day + "." + _month + "." + _year);
            }
        }, yy, mm, dd);
        datePicker.show();
    }

    public void selectPartners(View v) {
        final String[] partners = ((EditText) findViewById(R.id.partnersEditText)).getText().toString().split(", ");
        ArrayList<Integer> partnerIds = new ArrayList<Integer>();
        for (int i = 0; i < partners.length; i++) {
            partnerIds.add(_db.partnerDao().getId(partners[i]));
        }
        Intent intent = new Intent(AscendActivity.this, PartnersActivity.class);
        intent.putIntegerArrayListExtra(IntentConstants.ASCEND_PARTNER_IDS, partnerIds);
        startActivityForResult(intent, 0);
    }

    private void _displayContent() {
        setTitle(_route != null ? _route.getName() + "   " + _route.getGrade() : AppDatabase.UNKNOWN_NAME);


        Spinner spinner = findViewById(R.id.styleSpinner);
        ArrayAdapter<CharSequence> adapter = new ArrayAdapter<CharSequence>(this, android.R.layout.simple_list_item_1, AscendStyle.Companion.getNames());
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(adapter);
        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                _styleId = AscendStyle.Companion.fromName(parent.getItemAtPosition(position).toString()).getId();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
                _styleId = 0;
            }
        });

        if (_ascend != null) {
            ((EditText) findViewById(R.id.dateEditText)).setText((_day =_ascend.getDay()) + "." + (_month = _ascend.getMonth()) + "." + (_year = _ascend.getYear()));
            ((EditText) findViewById(R.id.notesEditText)).setText(_ascend.getNotes());
            spinner.setSelection(adapter.getPosition(AscendStyle.Companion.fromId(_styleId == 0 ? _ascend.getStyleId() : _styleId).getStyleName()));
        } else {
            spinner.setSelection(_styleId != 0 ? adapter.getPosition(AscendStyle.Companion.fromId(_styleId).getStyleName()) : 0);
        }

        ArrayList<String> partners = new ArrayList<String>();
        for (final Integer id : _partnerIds) {
            final Partner partner = _db.partnerDao().getPartner(id);
            partners.add(partner == null ? AppDatabase.UNKNOWN_NAME : partner.getName());
        }
        ((EditText) findViewById(R.id.partnersEditText)).setText(TextUtils.join(", ", partners));
    }
}

File: app/src/main/java/com/yacgroup/yacguide/AscendActivity.kt
/*
 * Copyright (C) 2019, 2022 Axel Paetzold
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.yacgroup.yacguide

import android.annotation.SuppressLint
import android.app.Activity
import android.app.DatePickerDialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import androidx.appcompat.app.AppCompatActivity
import android.text.TextUtils
import android.text.TextWatcher
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.*
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts

import com.yacgroup.yacguide.database.Ascend
import com.yacgroup.yacguide.database.DatabaseWrapper
import com.yacgroup.yacguide.database.Route
import com.yacgroup.yacguide.databinding.ActivityAscendBinding
import com.yacgroup.yacguide.utils.AscendStyle
import com.yacgroup.yacguide.utils.IntentConstants
import com.yacgroup.yacguide.utils.ParserUtils

import java.util.ArrayList
import java.util.Calendar

class AscendActivity : AppCompatActivity() {

    private lateinit var _binding: ActivityAscendBinding
    private lateinit var _db: DatabaseWrapper
    private lateinit var _ascend: Ascend
    private lateinit var _partnerResultLauncher: ActivityResultLauncher<Intent>
    private var _outdatedAscend: Ascend? = null
    private var _route: Route? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = ActivityAscendBinding.inflate(layoutInflater)
        setContentView(_binding.root)

        _db = DatabaseWrapper(this)

        // The database query is different from null,
        // if this activity is called from the TourbookAscendActivity (on editing an ascent).
        _outdatedAscend = _db.getAscend(intent.getIntExtra(IntentConstants.ASCEND_ID, DatabaseWrapper.INVALID_ID))
        // The intent constant CLIMBING_OBJECT_PARENT_ID is only available, if this activity is called from
        // the route DescriptionActivity (on entering a new ascent).
        _route = _db.getRoute(_outdatedAscend?.routeId
                    ?: intent.getIntExtra(IntentConstants.CLIMBING_OBJECT_PARENT_ID, DatabaseWrapper.INVALID_ID))
        // Beware: _route may still be null (if the route of this ascend has been deleted meanwhile)
        _ascend = Ascend(
            id = _outdatedAscend?.id ?: 0,
            routeId = _outdatedAscend?.routeId ?: _route!!.id,
            styleId = _outdatedAscend?.styleId ?: 0,
            year = _outdatedAscend?.year ?: 0,
            month = _outdatedAscend?.month ?: 0,
            day = _outdatedAscend?.day ?: 0,
            partnerIds = _outdatedAscend?.partnerIds ?: ArrayList(),
            notes = _outdatedAscend?.notes ?: ""
        )

        _partnerResultLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == Activity.RESULT_OK) {
                _ascend.partnerIds = result.data?.getIntegerArrayListExtra(IntentConstants.ASCEND_PARTNER_IDS)
                    ?: ArrayList()
            }
        }

        _binding.notesEditText.let {
            it.onFocusChangeListener = View.OnFocusChangeListener { view, _ ->
                val imm = view.context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                imm.hideSoftInputFromWindow(view.windowToken, 0)
            }
            it.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
                override fun afterTextChanged(s: Editable?) {
                    _ascend.notes = s.toString()
                }
            })
        }
    }

    public override fun onResume() {
        super.onResume()
        _displayContent()
    }

    @Suppress("UNUSED_PARAMETER")
    fun enter(v: View) {
        _outdatedAscend?.let { _db.deleteAscend(it) }
        _db.addAscend(_ascend)
        Toast.makeText(this, getString(R.string.ascends_refreshed), Toast.LENGTH_SHORT).show()
        finish()
    }

    @Suppress("UNUSED_PARAMETER")
    fun cancel(v: View) {
        finish()
    }

    @SuppressLint("SetTextI18n")
    @Suppress("UNUSED_PARAMETER")
    fun enterDate(v: View) {
        val calendar = Calendar.getInstance()
        val yy = _ascend.year.takeIf { it != 0 } ?: calendar.get(Calendar.YEAR)
        val mm = _ascend.month.takeIf { it != 0 }.let { it?.minus(1) } ?: calendar.get(Calendar.MONTH)
        val dd = _ascend.day.takeIf { it != 0 } ?: calendar.get(Calendar.DAY_OF_MONTH)
        val datePicker = DatePickerDialog(this@AscendActivity, { _, year, monthOfYear, dayOfMonth ->
            _ascend.year = year
            _ascend.month = monthOfYear + 1
            _ascend.day = dayOfMonth
            _binding.dateEditText.setText("${_ascend.day}.${_ascend.month}.${_ascend.year}")
        }, yy, mm, dd)
        datePicker.show()
    }

    @Suppress("UNUSED_PARAMETER")
    fun selectPartners(v: View) {
        val partnerNames = _binding.partnersEditText.text.toString().split(", ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val partnerIds = _db.getPartnerIds(partnerNames.toList()).filter { it > 0 } as ArrayList<Int>
        _partnerResultLauncher.launch(Intent(this@AscendActivity, PartnersActivity::class.java).apply {
            putIntegerArrayListExtra(IntentConstants.ASCEND_PARTNER_IDS, partnerIds)
        })
    }

    @SuppressLint("SetTextI18n")
    private fun _displayContent() {
        val routeName = ParserUtils.decodeObjectNames(_route?.name)
        this.title = "${if (routeName.first.isNotEmpty()) routeName.first else routeName.second} ${_route?.grade.orEmpty()}"

        val spinnerAdapter = ArrayAdapter<CharSequence>(this, R.layout.spinner_item, AscendStyle.names).apply {
            setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        }
        _binding.styleSpinner.apply {
            adapter = spinnerAdapter
            onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(
                    parent: AdapterView<*>,
                    view: View,
                    position: Int,
                    id: Long
                ) {
                    _ascend.styleId =
                        AscendStyle.fromName(parent.getItemAtPosition(position).toString())?.id
                            ?: AscendStyle.eUNKNOWN.id
                }

                override fun onNothingSelected(parent: AdapterView<*>) {
                    _ascend.styleId = AscendStyle.eUNKNOWN.id
                }
            }
            setSelection(spinnerAdapter.getPosition(AscendStyle.fromId(_ascend.styleId)?.styleName))
        }

        if (_ascend.day != 0 && _ascend.month != 0 && _ascend.year != 0) {
            _binding.dateEditText.setText("${_ascend.day}.${_ascend.month}.${_ascend.year}")
        }

        _binding.notesEditText.setText(_ascend.notes)

        val partnerNames = _db.getPartnerNames(_ascend.partnerIds.orEmpty().toList())
        _binding.partnersEditText.setText(TextUtils.join(", ", partnerNames))
    }
}
